source/labelrenderer.ts

```diff
@@ -0,0 +1,384 @@
+        gl.blendFunc(gl.SRC_ALPHA, gl.ONE_MINUS_SRC_ALPHA);
+
+        // gl.enable(gl.POLYGON_OFFSET_FILL);
+        // gl.polygonOffset(-1, 1); // avoid z-fighting with other geometry
+
+        this._program.bind();
+
+        const ndcOffset = this._ndcOffsetKernel.get(frameNumber);
+        ndcOffset[0] = 2.0 * ndcOffset[0] / this._frameSize[0];
+        ndcOffset[1] = 2.0 * ndcOffset[1] / this._frameSize[1];
+        gl.uniform2fv(this._uNdcOffset, ndcOffset);
+        gl.uniform1i(this._uFrameNumber, frameNumber);
+
+        gl.uniformMatrix4fv(this._uViewProjection, gl.GL_FALSE, this._camera.viewProjection);
+
+        this._fontFace.glyphTexture.bind(gl.TEXTURE0);
+        gl.uniform1i(this._uGlyphAtlas, 0);
+
+        this._intermediateFBO.clear(gl.COLOR_BUFFER_BIT, true, false);
+
+        this._labelGeometry.bind();
+        this._labelGeometry.draw();
+        this._labelGeometry.unbind();
+
+        this._intermediateFBO.unbind();
+
+        this._accumulate.frame(frameNumber);
+
+        this._fontFace.glyphTexture.unbind(gl.TEXTURE0);
+        gl.blendFunc(oldBlendSRC, oldBlendDST);
+        if (!wasBlendEnabled) {
+            gl.disable(gl.BLEND);
+        }
+
+        // gl.disable(gl.POLYGON_OFFSET_FILL);
+    }
+
+    protected onSwap(): void {
+        this._blit.framebuffer = this._accumulate.framebuffer ?
+            this._accumulate.framebuffer : this._blit.framebuffer = this._intermediateFBO;
+        this._blit.frame();
+    }
+
+    protected loadFont(context: Context): void {
+        const loader = new FontLoader();
+
+        const fontFace: FontFace = loader.load(
+            context, './data/opensansr144/opensansr144.fnt', false, () => {
+                this.setupScene();
+                this.invalidate();
+            });
+
+        this._fontFace = fontFace;
+    }
+
+    protected setupScene(): void {
+
+        // create Label with Text and
+        // tell the Typesetter to typeset that Label with the loaded FontFace
+
+        const userTransform = mat4.create();
+        mat4.scale(userTransform, userTransform, vec3.fromValues(1.2, 1.2, 1.2));
+        mat4.rotateX(userTransform, userTransform, Math.PI * -0.4);
+        mat4.rotateY(userTransform, userTransform, Math.PI * 0.2);
+        mat4.rotateZ(userTransform, userTransform, Math.PI * 0.5);
+        mat4.translate(userTransform, userTransform, vec3.fromValues(-1, 0.0, 0));
+        let glyphVertices = this.prepareLabel('Hello Transform!', userTransform);
+
+        glyphVertices = glyphVertices.concat(this.prepareLabel('Hello World!'));
+
+        const origins: Array<number> = [];
+        const tans: Array<number> = [];
+        const ups: Array<number> = [];
+        const texCoords: Array<number> = [];
+
+        const l = glyphVertices.length;
+
+        for (let i = 0; i < l; i++) {
+            // TODO: shouldn't there be an easier way to achieve this?
+            // concat doesn't work as vec3 apparently is not an Array.
```

Comment:
Yes - e.g. `origins.push(...v)`

Reply:
You mean `origins.push(v.origin)` or `origins.push(v.origin[0], v.origin[1], v.origin[2])`? The first one wouldn't push numbers but vec3; the second one is still ugly, but ugly on the same line. :D

Reply:
I meant `origins.push(...v.origin)`, though I just saw we'd need to adapt our TypeScript config to use the syntax (-> https://blog.mariusschulz.com/2017/06/30/typescript-2-3-downlevel-iteration-for-es3-es5). But the equivalent `origins.push.apply(origins, v.origin)` should work. See also https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax.

Reply:
done, thanks! :)